cmd/kubeadm/app/phases/selfhosting/selfhosting_volumes.go

```diff
@@ -0,0 +1,340 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package selfhosting
+
+import (
+	"fmt"
+	"io/ioutil"
+	"path"
+
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	kubeadmapi "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm"
+	kubeadmconstants "k8s.io/kubernetes/cmd/kubeadm/app/constants"
+)
+
+type tlsKeyPair struct {
+	name string
+	cert string
+	key  string
+}
+
+func k8sSelfHostedVolumeMount() v1.VolumeMount {
+	return v1.VolumeMount{
+		Name:      "k8s",
+		MountPath: kubeadmapi.GlobalEnvParams.KubernetesDir,
+		ReadOnly:  true,
+	}
+}
+
+func apiServerProjectedVolume(cfg *kubeadmapi.MasterConfiguration) v1.Volume {
+	return v1.Volume{
+		Name: "k8s",
+		VolumeSource: v1.VolumeSource{
+			Projected: &v1.ProjectedVolumeSource{
+				Sources: []v1.VolumeProjection{
+					{
+						Secret: &v1.SecretProjection{
+							LocalObjectReference: v1.LocalObjectReference{
+								Name: kubeadmconstants.CACertAndKeyBaseName,
+							},
+							Items: []v1.KeyToPath{
+								{
+									Key:  v1.TLSCertKey,
+									Path: path.Join(path.Base(cfg.CertificatesDir), kubeadmconstants.CACertName),
```

Comment:
We have a really problematic corner case here if you set `CertificatesDir` to `/var/lib/certs` for example, where the cert dir is not a subpath of k8s path.

I think we have to solve it by ALWAYS doing this when self-hosting:
 - /etc/kubernetes/*.conf for kubeconfig files
 - /etc/kubernetes/pki/*.crt|key for certs

But the source earlier might have been `/var/lib/certs` when using Static Pods; so we need to rewrite the necessary args to the fixed self-hosted paths.

This requires the ability to rewrite `.command`, which is coming in a PR from me today.